celery/canvas.py

```diff
@@ -1221,20 +1218,43 @@ def freeze(self, _id=None, group_id=None, chord=None,
             opts['group_index'] = group_index
         root_id = opts.setdefault('root_id', root_id)
         parent_id = opts.setdefault('parent_id', parent_id)
-        new_tasks = []
-        # Need to unroll subgroups early so that chord gets the
-        # right result instance for chord_unlock etc.
-        results = list(self._freeze_unroll(
-            new_tasks, group_id, chord, root_id, parent_id,
-        ))
-        if isinstance(self.tasks, MutableSequence):
-            self.tasks[:] = new_tasks
+        if isinstance(self.tasks, _regen):
+            tasks1, tasks2 = itertools.tee(self._unroll_tasks(self.tasks))
```

Comment:
Please add a comment explaining that we're draining from a generator.

Reply:
Added a comment.